### PR TITLE
BLS12-381 Crypto Primitives

### PR DESCRIPTION
**Pull request by mschoenebeck**
This is the corresponding `cdt` pull request to [1071](https://github.com/AntelopeIO/leap/pull/1071) of `leap`. Enable the following types and host functions for smart contracts.

Added the following types to `cdt`:
```
capi_bls_scalar  = uint8_t data[32];
capi_bls_fp      = uint8_t data[48];
capi_bls_fp2     = uint8_t data[96];
capi_bls_g1      = uint8_t data[144];
capi_bls_g2      = uint8_t data[288];
capi_bls_gt      = uint8_t data[576];
```

Added the following host functions to `cdt` (according to https://eips.ethereum.org/EIPS/eip-2537):
```
bls_g1_add(capi_bls_g1 op1, capi_bls_g1 op2, capi_bls_g1 result);
bls_g2_add(capi_bls_g2 op1, capi_bls_g2 op2, capi_bls_g2 result);
bls_g1_mul(capi_bls_g1 point, capi_bls_scalar scalar, bls_g1 result);
bls_g2_mul(capi_bls_g2 point, capi_bls_scalar scalar, bls_g2 result);
bls_g1_exp(capi_bls_g1[] points, capi_bls_scalar[] scalars, capi_bls_g1 result);
bls_g2_exp(capi_bls_g2[] points, capi_bls_scalar[] scalars, capi_bls_g2 result);
bls_pairing(capi_bls_g1[] g1_points, capi_bls_g2[] g2_points, capi_bls_gt result);
bls_g1_map(capi_bls_fp e, capi_bls_g1 result);
bls_g2_map(capi_bls_fp2 e, capi_bls_g2 result);
bls_fp_mod(std::array<uint8_t, 64> s, capi_bls_fp result);
```

Check out [mschoenebeck/aggsigtest](https://github.com/mschoenebeck/aggsigtest) for a sample contract using those new host functions.

### CODE REVIEW

**Review by dimas1185 — CHANGES_REQUESTED** [review]
please add integration tests.
need to do the following:
add test contract to the `tests/unit/test_contracts` that calls every bls function
add this test to the 'tests/integration/contracts.hpp.in'
add integration test to the 'tests/integration' that is using this contract to check that all actions work properly

File: libraries/eosiolib/core/eosio/crypto_bls_ext.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g2_map(const char* e, uint32_t e_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_fp_mod(const char* s, uint32_t s_len, char* res, uint32_t res_len);
+        }
+    }
+
+    int32_t bls_g1_add(const uint8_t* op1, const uint8_t* op2, uint8_t* res)
```
Comment (dimas1185) [inline libraries/eosiolib/core/eosio/crypto_bls_ext.hpp:46]: why did you add those `uint8_t` overloaded versions?

Comment (dimas1185) [inline libraries/eosiolib/core/eosio/crypto_bls_ext.hpp:46]: I think it is better to have something like this:
```c++
int32_t bls_g1_add(const uint8_t (&op1)[144], const (&op2)[144], uint8_t (&res)[144])
```
also, you could typedef those and it will be even nicer:
```c++
using some_type = uint8_t[144];
    int32_t bls_g1_add(const some_type& op1, const some_type& op2, some_type& res)
```

**Comment by mschoenebeck** [comment]
Added integration tests for all bls primitives in: https://github.com/mschoenebeck/cdt/commit/7a6dfcb4a5165d31514c5e20648a01f3c700fea0

File: libraries/eosiolib/core/eosio/crypto_bls_ext.hpp (hunk 1 above)
Comment (mschoenebeck) [reply libraries/eosiolib/core/eosio/crypto_bls_ext.hpp]: Added bls types to enforce correct operand sizes in: https://github.com/mschoenebeck/cdt/commit/9d1424dfcf038cbaa096dda1976fcc44c0443adb

**Review by dimas1185 — APPROVED** [review]

**Comment by spoonincode** [comment]
The integration tests are not passing because the `leap-dev.deb` being used is from 4.0, which of course will not have support for the new bls host functions,
https://github.com/AntelopeIO/cdt/blob/3f701e2ea9c3a81d5030bbe0d08caf86ec0c969c/.github/workflows/build.yaml#L70-L80
If we'd like to see the tests pass here (and continue to pass post merge), we need to switch the `leap-dev.deb` over to one from leap's BLS PR. Sadly since it's a fork, referencing it by branch won't work, but I'm pretty sure using
```yaml
   target: 90071617799738820350538f632ca5d5fd54e5e8
```
will work (this is the head of AntelopeIO/leap#1071 at the moment)

Once AntelopeIO/leap#1071 is merged in to leap's main, the above will need to become
```yaml
   target: main
```
until 5.0 is cut, at which point it can then be changed to
```yaml
   target: 5
```

**Review by arhag — APPROVED** [review]
Overall this looks good to me. There are some small tweaks the ENF will do including getting the CI to fully pass. But we are good as it is to merge this into the `bls_integraton` branch.
